### DISCUSSION TITLE
PROPOSAL:  Sorting of Unit Tests

### DISCUSSION THREAD

**Discussion by cjdutoit** [hassanhabib/The-Standard#227] [Ideas]
I would like to propose a change in the naming convention of unit tests.

We currently use the convention of:
`%Service%`Tests.Exceptions.`%Method%`.cs
`%Service%`Tests.Logic.`%Method%`.cs
`%Service%`Tests.Validations.`%Method%`.cs

I would like to suggest that we change this to the following convention:
`%Service%`Tests.`%Method%`.Exceptions.cs
`%Service%`Tests.`%Method%`.Logic.cs
`%Service%`Tests.`%Method%`.Validations.cs

The proposed convention will logically group the tests together per method on disk / Solution Explorer, making it easy to see and navigate.  If we consider all the types off tests, the proposed result could look like this:

```cs
    StudentServiceTests.Add.Exceptions.cs
    StudentServiceTests.Add.Logic.cs
    StudentServiceTests.Add.Rollbacks.cs
    StudentServiceTests.Add.Securities.cs
    StudentServiceTests.Add.Validations.cs
    StudentServiceTests.Modify.Exceptions.cs
    StudentServiceTests.Modify.Logic.cs
    StudentServiceTests.Modify.Rollbacks.cs
    StudentServiceTests.Modify.Securities.cs
    StudentServiceTests.Modify.Validations.cs
```
instead of 
```cs
    StudentServiceTests.Exceptions.Add.cs
    StudentServiceTests.Exceptions.Modify.cs
    StudentServiceTests.Logics.Add.cs
    StudentServiceTests.Logics.Modify.cs
    StudentServiceTests.Rollbacks.Add.cs
    StudentServiceTests.Rollbacks.Modify.cs
    StudentServiceTests.Securities.Add.cs
    StudentServiceTests.Securities.Modify.cs
    StudentServiceTests.Validations.Add.cs
    StudentServiceTests.Validations.Modify.cs
```

(As the files is sorted alphabetically, we do not quite see the exact order in which we do TDD, so I think it will be more beneficial to see the grouping by method)

**Comment by hassanhabib** [comment]
@cjdutoit my only argument on this is that you want to work with the operation as a full unit. So when I'm developing an `AddStudent` feature I want to see all `Add` aspects under each other.
In other words, I haven't yet found the need to look for all exceptions across all operations in one Service to make a change. It's always Logic -> Validations -> Exceptions.

**Reply by cjdutoit** [reply]
@hassanhabib, are we not saying the same thing then with my suggested syntax?

Moving the operation forward in the name convention will effectively group all the tests together by operation like this:
(service, then operation, then exception type of the test)

![image](https://user-images.githubusercontent.com/797750/211775438-188cf8f5-dd97-4bb0-9317-37968fea0877.png)


The current convention groups, by service, then exception then operation which scatter the tests for an `Add`.


![image](https://user-images.githubusercontent.com/797750/211776890-3c766428-b111-4294-abb5-ad624d83dff0.png)

(I get what you say about the order of Logic -> Validations -> Exceptions, but the alphabetical sorting on the filenames does not give us that sorting /benefit at the moment either.)
